HOROVOD_WITH_PYTORCH=1 HOROVOD_WITHOUT_TENSORFLOW=1 pip3 install horovod[pytorch] install failed

HOROVOD_WITH_PYTORCH=1 HOROVOD_WITHOUT_TENSORFLOW=1 pip3 install horovod[pytorch]

**Environment:**
1. Framework:pytorch
2. Framework version:1.9.0
3. Horovod version:
4. MPI version:
5. CUDA version:11.1
6. NCCL version:
7. Python version:3.8
8. Spark / PySpark version:
9. Ray version:
10. OS and version:ubuntu20.04
11. GCC version:
12. CMake version:

```
    ERROR: Command errored out with exit status 1:
     command: /usr/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-o6yhlor8/horovod/setup.py'"'"'; __file__='"'"'/tmp/pip-install-o6yhlor8/horovod/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-zhlcztka/install-record.txt --single-version-externally-managed --user --prefix= --compile --install-headers /home/sczone/.local/include/python3.8/horovod
         cwd: /tmp/pip-install-o6yhlor8/horovod/
    Complete output (346 lines):
    running install
    running build
    running build_py
    creating build
    creating build/lib.linux-x86_64-3.8
    creating build/lib.linux-x86_64-3.8/horovod
    copying horovod/__init__.py -> build/lib.linux-x86_64-3.8/horovod
    creating build/lib.linux-x86_64-3.8/horovod/mxnet
    copying horovod/mxnet/__init__.py -> build/lib.linux-x86_64-3.8/horovod/mxnet
    copying horovod/mxnet/compression.py -> build/lib.linux-x86_64-3.8/horovod/mxnet
    copying horovod/mxnet/functions.py -> build/lib.linux-x86_64-3.8/horovod/mxnet
    copying horovod/mxnet/mpi_ops.py -> build/lib.linux-x86_64-3.8/horovod/mxnet
    creating build/lib.linux-x86_64-3.8/horovod/spark
    copying horovod/spark/conf.py -> build/lib.linux-x86_64-3.8/horovod/spark
    copying horovod/spark/mpi_run.py -> build/lib.linux-x86_64-3.8/horovod/spark
    copying horovod/spark/__init__.py -> build/lib.linux-x86_64-3.8/horovod/spark
    copying horovod/spark/gloo_run.py -> build/lib.linux-x86_64-3.8/horovod/spark
    copying horovod/spark/runner.py -> build/lib.linux-x86_64-3.8/horovod/spark
    creating build/lib.linux-x86_64-3.8/horovod/torch
    copying horovod/torch/__init__.py -> build/lib.linux-x86_64-3.8/horovod/torch
    copying horovod/torch/compression.py -> build/lib.linux-x86_64-3.8/horovod/torch
    copying horovod/torch/functions.py -> build/lib.linux-x86_64-3.8/horovod/torch
    copying horovod/torch/mpi_ops.py -> build/lib.linux-x86_64-3.8/horovod/torch
    copying horovod/torch/sync_batch_norm.py -> build/lib.linux-x86_64-3.8/horovod/torch
    copying horovod/torch/optimizer.py -> build/lib.linux-x86_64-3.8/horovod/torch
    creating build/lib.linux-x86_64-3.8/horovod/_keras
    copying horovod/_keras/__init__.py -> build/lib.linux-x86_64-3.8/horovod/_keras
    copying horovod/_keras/callbacks.py -> build/lib.linux-x86_64-3.8/horovod/_keras
    copying horovod/_keras/elastic.py -> build/lib.linux-x86_64-3.8/horovod/_keras
    creating build/lib.linux-x86_64-3.8/horovod/runner
    copying horovod/runner/mpi_run.py -> build/lib.linux-x86_64-3.8/horovod/runner
    copying horovod/runner/__init__.py -> build/lib.linux-x86_64-3.8/horovod/runner
    copying horovod/runner/gloo_run.py -> build/lib.linux-x86_64-3.8/horovod/runner
    copying horovod/runner/run_task.py -> build/lib.linux-x86_64-3.8/horovod/runner
    copying horovod/runner/js_run.py -> build/lib.linux-x86_64-3.8/horovod/runner
    copying horovod/runner/launch.py -> build/lib.linux-x86_64-3.8/horovod/runner
    copying horovod/runner/task_fn.py -> build/lib.linux-x86_64-3.8/horovod/runner
    creating build/lib.linux-x86_64-3.8/horovod/data
    copying horovod/data/__init__.py -> build/lib.linux-x86_64-3.8/horovod/data
    copying horovod/data/data_loader_base.py -> build/lib.linux-x86_64-3.8/horovod/data
    creating build/lib.linux-x86_64-3.8/horovod/tensorflow
    copying horovod/tensorflow/__init__.py -> build/lib.linux-x86_64-3.8/horovod/tensorflow
    copying horovod/tensorflow/compression.py -> build/lib.linux-x86_64-3.8/horovod/tensorflow
    copying horovod/tensorflow/gradient_aggregation.py -> build/lib.linux-x86_64-3.8/horovod/tensorflow
    copying horovod/tensorflow/functions.py -> build/lib.linux-x86_64-3.8/horovod/tensorflow
    copying horovod/tensorflow/gradient_aggregation_eager.py -> build/lib.linux-x86_64-3.8/horovod/tensorflow
    copying horovod/tensorflow/mpi_ops.py -> build/lib.linux-x86_64-3.8/horovod/tensorflow
    copying horovod/tensorflow/sync_batch_norm.py -> build/lib.linux-x86_64-3.8/horovod/tensorflow
    copying horovod/tensorflow/util.py -> build/lib.linux-x86_64-3.8/horovod/tensorflow
    copying horovod/tensorflow/elastic.py -> build/lib.linux-x86_64-3.8/horovod/tensorflow
    creating build/lib.linux-x86_64-3.8/horovod/ray
    copying horovod/ray/utils.py -> build/lib.linux-x86_64-3.8/horovod/ray
    copying horovod/ray/ray_logger.py -> build/lib.linux-x86_64-3.8/horovod/ray
    copying horovod/ray/driver_service.py -> build/lib.linux-x86_64-3.8/horovod/ray
    copying horovod/ray/__init__.py -> build/lib.linux-x86_64-3.8/horovod/ray
    copying horovod/ray/elastic_v2.py -> build/lib.linux-x86_64-3.8/horovod/ray
    copying horovod/ray/worker.py -> build/lib.linux-x86_64-3.8/horovod/ray
    copying horovod/ray/runner.py -> build/lib.linux-x86_64-3.8/horovod/ray
    copying horovod/ray/strategy.py -> build/lib.linux-x86_64-3.8/horovod/ray
    copying horovod/ray/adapter.py -> build/lib.linux-x86_64-3.8/horovod/ray
    copying horovod/ray/elastic.py -> build/lib.linux-x86_64-3.8/horovod/ray
    creating build/lib.linux-x86_64-3.8/horovod/keras
    copying horovod/keras/__init__.py -> build/lib.linux-x86_64-3.8/horovod/keras
    copying horovod/keras/callbacks.py -> build/lib.linux-x86_64-3.8/horovod/keras
    copying horovod/keras/elastic.py -> build/lib.linux-x86_64-3.8/horovod/keras
    creating build/lib.linux-x86_64-3.8/horovod/common
    copying horovod/common/exceptions.py -> build/lib.linux-x86_64-3.8/horovod/common
    copying horovod/common/__init__.py -> build/lib.linux-x86_64-3.8/horovod/common
    copying horovod/common/basics.py -> build/lib.linux-x86_64-3.8/horovod/common
    copying horovod/common/util.py -> build/lib.linux-x86_64-3.8/horovod/common
    copying horovod/common/elastic.py -> build/lib.linux-x86_64-3.8/horovod/common
    copying horovod/common/process_sets.py -> build/lib.linux-x86_64-3.8/horovod/common
    creating build/lib.linux-x86_64-3.8/horovod/spark/driver
    copying horovod/spark/driver/rsh.py -> build/lib.linux-x86_64-3.8/horovod/spark/driver
    copying horovod/spark/driver/mpirun_rsh.py -> build/lib.linux-x86_64-3.8/horovod/spark/driver
    copying horovod/spark/driver/driver_service.py -> build/lib.linux-x86_64-3.8/horovod/spark/driver
    copying horovod/spark/driver/__init__.py -> build/lib.linux-x86_64-3.8/horovod/spark/driver
    copying horovod/spark/driver/job_id.py -> build/lib.linux-x86_64-3.8/horovod/spark/driver
    copying horovod/spark/driver/host_discovery.py -> build/lib.linux-x86_64-3.8/horovod/spark/driver
    copying horovod/spark/driver/rendezvous.py -> build/lib.linux-x86_64-3.8/horovod/spark/driver
    creating build/lib.linux-x86_64-3.8/horovod/spark/torch
    copying horovod/spark/torch/remote.py -> build/lib.linux-x86_64-3.8/horovod/spark/torch
    copying horovod/spark/torch/estimator.py -> build/lib.linux-x86_64-3.8/horovod/spark/torch
    copying horovod/spark/torch/__init__.py -> build/lib.linux-x86_64-3.8/horovod/spark/torch
    copying horovod/spark/torch/util.py -> build/lib.linux-x86_64-3.8/horovod/spark/torch
    creating build/lib.linux-x86_64-3.8/horovod/spark/data_loaders
    copying horovod/spark/data_loaders/__init__.py -> build/lib.linux-x86_64-3.8/horovod/spark/data_loaders
    copying horovod/spark/data_loaders/pytorch_data_loaders.py -> build/lib.linux-x86_64-3.8/horovod/spark/data_loaders
    creating build/lib.linux-x86_64-3.8/horovod/spark/task
    copying horovod/spark/task/__init__.py -> build/lib.linux-x86_64-3.8/horovod/spark/task
    copying horovod/spark/task/task_service.py -> build/lib.linux-x86_64-3.8/horovod/spark/task
    copying horovod/spark/task/gloo_exec_fn.py -> build/lib.linux-x86_64-3.8/horovod/spark/task
    copying horovod/spark/task/task_info.py -> build/lib.linux-x86_64-3.8/horovod/spark/task
    copying horovod/spark/task/mpirun_exec_fn.py -> build/lib.linux-x86_64-3.8/horovod/spark/task
    creating build/lib.linux-x86_64-3.8/horovod/spark/keras
    copying horovod/spark/keras/remote.py -> build/lib.linux-x86_64-3.8/horovod/spark/keras
    copying horovod/spark/keras/tensorflow.py -> build/lib.linux-x86_64-3.8/horovod/spark/keras
    copying horovod/spark/keras/estimator.py -> build/lib.linux-x86_64-3.8/horovod/spark/keras
    copying horovod/spark/keras/__init__.py -> build/lib.linux-x86_64-3.8/horovod/spark/keras
    copying horovod/spark/keras/bare.py -> build/lib.linux-x86_64-3.8/horovod/spark/keras
    copying horovod/spark/keras/optimizer.py -> build/lib.linux-x86_64-3.8/horovod/spark/keras
    copying horovod/spark/keras/util.py -> build/lib.linux-x86_64-3.8/horovod/spark/keras
    creating build/lib.linux-x86_64-3.8/horovod/spark/common
    copying horovod/spark/common/serialization.py -> build/lib.linux-x86_64-3.8/horovod/spark/common
    copying horovod/spark/common/estimator.py -> build/lib.linux-x86_64-3.8/horovod/spark/common
    copying horovod/spark/common/__init__.py -> build/lib.linux-x86_64-3.8/horovod/spark/common
    copying horovod/spark/common/_namedtuple_fix.py -> build/lib.linux-x86_64-3.8/horovod/spark/common
    copying horovod/spark/common/cache.py -> build/lib.linux-x86_64-3.8/horovod/spark/common
    copying horovod/spark/common/store.py -> build/lib.linux-x86_64-3.8/horovod/spark/common
    copying horovod/spark/common/params.py -> build/lib.linux-x86_64-3.8/horovod/spark/common
    copying horovod/spark/common/constants.py -> build/lib.linux-x86_64-3.8/horovod/spark/common
    copying horovod/spark/common/util.py -> build/lib.linux-x86_64-3.8/horovod/spark/common
    copying horovod/spark/common/backend.py -> build/lib.linux-x86_64-3.8/horovod/spark/common
    creating build/lib.linux-x86_64-3.8/horovod/spark/lightning
    copying horovod/spark/lightning/remote.py -> build/lib.linux-x86_64-3.8/horovod/spark/lightning
    copying horovod/spark/lightning/legacy.py -> build/lib.linux-x86_64-3.8/horovod/spark/lightning
    copying horovod/spark/lightning/estimator.py -> build/lib.linux-x86_64-3.8/horovod/spark/lightning
    copying horovod/spark/lightning/__init__.py -> build/lib.linux-x86_64-3.8/horovod/spark/lightning
    copying horovod/spark/lightning/datamodule.py -> build/lib.linux-x86_64-3.8/horovod/spark/lightning
    copying horovod/spark/lightning/util.py -> build/lib.linux-x86_64-3.8/horovod/spark/lightning
    creating build/lib.linux-x86_64-3.8/horovod/torch/mpi_lib_impl
    copying horovod/torch/mpi_lib_impl/__init__.py -> build/lib.linux-x86_64-3.8/horovod/torch/mpi_lib_impl
    creating build/lib.linux-x86_64-3.8/horovod/torch/mpi_lib
    copying horovod/torch/mpi_lib/__init__.py -> build/lib.linux-x86_64-3.8/horovod/torch/mpi_lib
    creating build/lib.linux-x86_64-3.8/horovod/torch/elastic
    copying horovod/torch/elastic/state.py -> build/lib.linux-x86_64-3.8/horovod/torch/elastic
    copying horovod/torch/elastic/__init__.py -> build/lib.linux-x86_64-3.8/horovod/torch/elastic
    copying horovod/torch/elastic/sampler.py -> build/lib.linux-x86_64-3.8/horovod/torch/elastic
    creating build/lib.linux-x86_64-3.8/horovod/runner/driver
    copying horovod/runner/driver/driver_service.py -> build/lib.linux-x86_64-3.8/horovod/runner/driver
    copying horovod/runner/driver/__init__.py -> build/lib.linux-x86_64-3.8/horovod/runner/driver
    creating build/lib.linux-x86_64-3.8/horovod/runner/elastic
    copying horovod/runner/elastic/settings.py -> build/lib.linux-x86_64-3.8/horovod/runner/elastic
    copying horovod/runner/elastic/driver.py -> build/lib.linux-x86_64-3.8/horovod/runner/elastic
    copying horovod/runner/elastic/__init__.py -> build/lib.linux-x86_64-3.8/horovod/runner/elastic
    copying horovod/runner/elastic/worker.py -> build/lib.linux-x86_64-3.8/horovod/runner/elastic
    copying horovod/runner/elastic/discovery.py -> build/lib.linux-x86_64-3.8/horovod/runner/elastic
    copying horovod/runner/elastic/constants.py -> build/lib.linux-x86_64-3.8/horovod/runner/elastic
    copying horovod/runner/elastic/rendezvous.py -> build/lib.linux-x86_64-3.8/horovod/runner/elastic
    copying horovod/runner/elastic/registration.py -> build/lib.linux-x86_64-3.8/horovod/runner/elastic
    creating build/lib.linux-x86_64-3.8/horovod/runner/task
    copying horovod/runner/task/__init__.py -> build/lib.linux-x86_64-3.8/horovod/runner/task
    copying horovod/runner/task/task_service.py -> build/lib.linux-x86_64-3.8/horovod/runner/task
    creating build/lib.linux-x86_64-3.8/horovod/runner/http
    copying horovod/runner/http/http_server.py -> build/lib.linux-x86_64-3.8/horovod/runner/http
    copying horovod/runner/http/http_client.py -> build/lib.linux-x86_64-3.8/horovod/runner/http
    copying horovod/runner/http/__init__.py -> build/lib.linux-x86_64-3.8/horovod/runner/http
    creating build/lib.linux-x86_64-3.8/horovod/runner/util
    copying horovod/runner/util/remote.py -> build/lib.linux-x86_64-3.8/horovod/runner/util
    copying horovod/runner/util/lsf.py -> build/lib.linux-x86_64-3.8/horovod/runner/util
    copying horovod/runner/util/__init__.py -> build/lib.linux-x86_64-3.8/horovod/runner/util
    copying horovod/runner/util/cache.py -> build/lib.linux-x86_64-3.8/horovod/runner/util
    copying horovod/runner/util/streams.py -> build/lib.linux-x86_64-3.8/horovod/runner/util
    copying horovod/runner/util/network.py -> build/lib.linux-x86_64-3.8/horovod/runner/util
    copying horovod/runner/util/threads.py -> build/lib.linux-x86_64-3.8/horovod/runner/util
    creating build/lib.linux-x86_64-3.8/horovod/runner/common
    copying horovod/runner/common/__init__.py -> build/lib.linux-x86_64-3.8/horovod/runner/common
    creating build/lib.linux-x86_64-3.8/horovod/runner/common/service
    copying horovod/runner/common/service/driver_service.py -> build/lib.linux-x86_64-3.8/horovod/runner/common/service
    copying horovod/runner/common/service/__init__.py -> build/lib.linux-x86_64-3.8/horovod/runner/common/service
    copying horovod/runner/common/service/task_service.py -> build/lib.linux-x86_64-3.8/horovod/runner/common/service
    creating build/lib.linux-x86_64-3.8/horovod/runner/common/util
    copying horovod/runner/common/util/codec.py -> build/lib.linux-x86_64-3.8/horovod/runner/common/util
    copying horovod/runner/common/util/settings.py -> build/lib.linux-x86_64-3.8/horovod/runner/common/util
    copying horovod/runner/common/util/secret.py -> build/lib.linux-x86_64-3.8/horovod/runner/common/util
    copying horovod/runner/common/util/__init__.py -> build/lib.linux-x86_64-3.8/horovod/runner/common/util
    copying horovod/runner/common/util/safe_shell_exec.py -> build/lib.linux-x86_64-3.8/horovod/runner/common/util
    copying horovod/runner/common/util/config_parser.py -> build/lib.linux-x86_64-3.8/horovod/runner/common/util
    copying horovod/runner/common/util/timeout.py -> build/lib.linux-x86_64-3.8/horovod/runner/common/util
    copying horovod/runner/common/util/env.py -> build/lib.linux-x86_64-3.8/horovod/runner/common/util
    copying horovod/runner/common/util/host_hash.py -> build/lib.linux-x86_64-3.8/horovod/runner/common/util
    copying horovod/runner/common/util/network.py -> build/lib.linux-x86_64-3.8/horovod/runner/common/util
    copying horovod/runner/common/util/hosts.py -> build/lib.linux-x86_64-3.8/horovod/runner/common/util
    copying horovod/runner/common/util/tiny_shell_exec.py -> build/lib.linux-x86_64-3.8/horovod/runner/common/util
    creating build/lib.linux-x86_64-3.8/horovod/tensorflow/keras
    copying horovod/tensorflow/keras/__init__.py -> build/lib.linux-x86_64-3.8/horovod/tensorflow/keras
    copying horovod/tensorflow/keras/callbacks.py -> build/lib.linux-x86_64-3.8/horovod/tensorflow/keras
    copying horovod/tensorflow/keras/elastic.py -> build/lib.linux-x86_64-3.8/horovod/tensorflow/keras
    running build_ext
    Running CMake in build/temp.linux-x86_64-3.8/RelWithDebInfo:
    cmake /tmp/pip-install-o6yhlor8/horovod -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_LIBRARY_OUTPUT_DIRECTORY_RELWITHDEBINFO=/tmp/pip-install-o6yhlor8/horovod/build/lib.linux-x86_64-3.8 -DPYTHON_EXECUTABLE:FILEPATH=/usr/bin/python3
    cmake --build . --config RelWithDebInfo -- -j8 VERBOSE=1
    -- Could not find CCache. Consider installing CCache to speed up compilation.
    -- The CXX compiler identification is GNU 9.3.0
    -- Check for working CXX compiler: /usr/bin/c++
    -- Check for working CXX compiler: /usr/bin/c++ -- works
    -- Detecting CXX compiler ABI info
    -- Detecting CXX compiler ABI info - done
    -- Detecting CXX compile features
    -- Detecting CXX compile features - done
    -- Build architecture flags: -mf16c -mavx -mfma
    -- Using command /usr/bin/python3
    -- Found MPI_CXX: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so (found version "3.1")
    -- Found MPI: TRUE (found version "3.1")
    -- Looking for a CUDA compiler
    -- Looking for a CUDA compiler - NOTFOUND
    -- Looking for a CUDA host compiler - /usr/bin/c++
    -- Found CUDAToolkit: /usr/local/cuda/include (found version "11.1.74")
    -- Looking for C++ include pthread.h
    -- Looking for C++ include pthread.h - found
    -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
    -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
    -- Found Threads: TRUE
    CUDA compiler was not found in $PATH, but searching again in CUDA Toolkit binary directory
    -- The CUDA compiler identification is NVIDIA 11.1.74
    -- Check for working CUDA compiler: /usr/local/cuda/bin/nvcc
    -- Check for working CUDA compiler: /usr/local/cuda/bin/nvcc -- works
    -- Detecting CUDA compiler ABI info
    -- Detecting CUDA compiler ABI info - done
    -- Found NVTX: /usr/local/cuda/include
    -- Found NVTX (include: /usr/local/cuda/include, library: dl)
    -- The C compiler identification is GNU 9.3.0
    -- Check for working C compiler: /usr/bin/cc
    -- Check for working C compiler: /usr/bin/cc -- works
    -- Detecting C compiler ABI info
    -- Detecting C compiler ABI info - done
    -- Detecting C compile features
    -- Detecting C compile features - done
    -- Gloo build as STATIC library
    -- Found MPI_C: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so (found version "3.1")
    -- Found MPI: TRUE (found version "3.1")
    -- MPI include path: /usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/usr/lib/x86_64-linux-gnu/openmpi/include
    -- MPI libraries: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so/usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so
    -- Found Pytorch: 1.9.1+cu111 (found suitable version "1.9.1+cu111", minimum required is "1.2.0")
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
    ModuleNotFoundError: No module named 'mxnet'
    -- Could NOT find Mxnet (missing: Mxnet_LIBRARIES) (Required is at least version "1.4.0")
    -- HVD_NVCC_COMPILE_FLAGS = -O3 -Xcompiler -fPIC -gencode arch=compute_35,code=sm_35 -gencode arch=compute_37,code=sm_37 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_52,code=sm_52 -gencode arch=compute_53,code=sm_53 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_62,code=sm_62 -gencode arch=compute_70,code=sm_70 -gencode arch=compute_72,code=sm_72 -gencode arch=compute_75,code=sm_75 -gencode arch=compute_80,code=sm_80 -gencode arch=compute_86,code=sm_86 -gencode arch=compute_86,code=compute_86
    -- Gloo build as STATIC library
    -- MPI include path: /usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/usr/lib/x86_64-linux-gnu/openmpi/include
    -- MPI libraries: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so/usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so
    -- Configuring done
    You have changed variables that require your cache to be deleted.
    Configure will be re-run and you may have to reset some variables.
    The following variables have changed:
    CMAKE_CUDA_COMPILER= NOTFOUND
    
    -- Could not find CCache. Consider installing CCache to speed up compilation.
    -- The CXX compiler identification is GNU 9.3.0
    -- Check for working CXX compiler: /usr/bin/c++
    -- Check for working CXX compiler: /usr/bin/c++ -- works
    -- Detecting CXX compiler ABI info
    -- Detecting CXX compiler ABI info - done
    -- Detecting CXX compile features
    -- Detecting CXX compile features - done
    -- Build architecture flags: -mf16c -mavx -mfma
    -- Found Python: /usr/bin/python3.8 (found suitable version "3.8.10", minimum required is "3.6") found components: Interpreter
    -- Using command /usr/bin/python3.8
    -- Found MPI_CXX: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so (found version "3.1")
    -- Found MPI: TRUE (found version "3.1")
    -- Found CUDAToolkit: /usr/local/cuda/include (found version "11.1.74")
    -- Looking for C++ include pthread.h
    -- Looking for C++ include pthread.h - found
    -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
    -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
    -- Found Threads: TRUE
    CUDA compiler was not found in $PATH, but searching again in CUDA Toolkit binary directory
    -- The CUDA compiler identification is NVIDIA 11.1.74
    -- Check for working CUDA compiler: /usr/local/cuda/bin/nvcc
    -- Check for working CUDA compiler: /usr/local/cuda/bin/nvcc -- works
    -- Detecting CUDA compiler ABI info
    -- Detecting CUDA compiler ABI info - done
    -- Found NVTX: /usr/local/cuda/include
    -- Found NVTX (include: /usr/local/cuda/include, library: dl)
    -- The C compiler identification is GNU 9.3.0
    -- Check for working C compiler: /usr/bin/cc
    -- Check for working C compiler: /usr/bin/cc -- works
    -- Detecting C compiler ABI info
    -- Detecting C compiler ABI info - done
    -- Detecting C compile features
    -- Detecting C compile features - done
    -- Build type not set -- defaulting to Release
    -- Gloo build as STATIC library
    -- Found MPI_C: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so (found version "3.1")
    -- Found MPI: TRUE (found version "3.1")
    -- MPI include path: /usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/usr/lib/x86_64-linux-gnu/openmpi/include
    -- MPI libraries: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so/usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so
    -- Found Pytorch: 1.9.1+cu111 (found suitable version "1.9.1+cu111", minimum required is "1.2.0")
    CMake Error at horovod/torch/CMakeLists.txt:18 (file):
      file failed to open for writing (Permission denied):
    
        /metadata.json
    
    
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
    ModuleNotFoundError: No module named 'mxnet'
    -- Could NOT find Mxnet (missing: Mxnet_LIBRARIES) (Required is at least version "1.4.0")
    CMake Error at CMakeLists.txt:345 (file):
      file failed to open for writing (Permission denied):
    
        /metadata.json
    
    
    -- HVD_NVCC_COMPILE_FLAGS = -O3 -Xcompiler -fPIC -gencode arch=compute_35,code=sm_35 -gencode arch=compute_37,code=sm_37 -gencode arch=compute_50,code=sm_50 -gencode arch=compute_52,code=sm_52 -gencode arch=compute_53,code=sm_53 -gencode arch=compute_60,code=sm_60 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_62,code=sm_62 -gencode arch=compute_70,code=sm_70 -gencode arch=compute_72,code=sm_72 -gencode arch=compute_75,code=sm_75 -gencode arch=compute_80,code=sm_80 -gencode arch=compute_86,code=sm_86 -gencode arch=compute_86,code=compute_86
    -- Build type not set -- defaulting to Release
    -- Gloo build as STATIC library
    -- MPI include path: /usr/lib/x86_64-linux-gnu/openmpi/include/openmpi/usr/lib/x86_64-linux-gnu/openmpi/include
    -- MPI libraries: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so/usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so
    -- Configuring incomplete, errors occurred!
    See also "/tmp/pip-install-o6yhlor8/horovod/build/temp.linux-x86_64-3.8/RelWithDebInfo/CMakeFiles/CMakeOutput.log".
    You have changed variables that require your cache to be deleted.
    Configure will be re-run and you may have to reset some variables.
    The following variables have changed:
    CMAKE_CUDA_COMPILER= NOTFOUND
    
    -- Generating done
    CMake Generate step failed.  Build files cannot be regenerated correctly.
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-install-o6yhlor8/horovod/setup.py", line 209, in <module>
        setup(name='horovod',
      File "/usr/lib/python3/dist-packages/setuptools/__init__.py", line 144, in setup
        return distutils.core.setup(**attrs)
      File "/usr/lib/python3.8/distutils/core.py", line 148, in setup
        dist.run_commands()
      File "/usr/lib/python3.8/distutils/dist.py", line 966, in run_commands
        self.run_command(cmd)
      File "/usr/lib/python3.8/distutils/dist.py", line 985, in run_command
        cmd_obj.run()
      File "/usr/lib/python3/dist-packages/setuptools/command/install.py", line 61, in run
        return orig.install.run(self)
      File "/usr/lib/python3.8/distutils/command/install.py", line 589, in run
        self.run_command('build')
      File "/usr/lib/python3.8/distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/usr/lib/python3.8/distutils/dist.py", line 985, in run_command
        cmd_obj.run()
      File "/usr/lib/python3.8/distutils/command/build.py", line 135, in run
        self.run_command(cmd_name)
      File "/usr/lib/python3.8/distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/usr/lib/python3.8/distutils/dist.py", line 985, in run_command
        cmd_obj.run()
      File "/usr/lib/python3/dist-packages/setuptools/command/build_ext.py", line 87, in run
        _build_ext.run(self)
      File "/home/sczone/.local/lib/python3.8/site-packages/Cython/Distutils/old_build_ext.py", line 186, in run
        _build_ext.build_ext.run(self)
      File "/usr/lib/python3.8/distutils/command/build_ext.py", line 340, in run
        self.build_extensions()
      File "/tmp/pip-install-o6yhlor8/horovod/setup.py", line 144, in build_extensions
        subprocess.check_call(command, cwd=cmake_build_dir)
      File "/usr/lib/python3.8/subprocess.py", line 364, in check_call
        raise CalledProcessError(retcode, cmd)
    subprocess.CalledProcessError: Command '['cmake', '/tmp/pip-install-o6yhlor8/horovod', '-DCMAKE_BUILD_TYPE=RelWithDebInfo', '-DCMAKE_LIBRARY_OUTPUT_DIRECTORY_RELWITHDEBINFO=/tmp/pip-install-o6yhlor8/horovod/build/lib.linux-x86_64-3.8', '-DPYTHON_EXECUTABLE:FILEPATH=/usr/bin/python3']' returned non-zero exit status 1.
    ----------------------------------------
ERROR: Command errored out with exit status 1: /usr/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-o6yhlor8/horovod/setup.py'"'"'; __file__='"'"'/tmp/pip-install-o6yhlor8/horovod/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-zhlcztka/install-record.txt --single-version-externally-managed --user --prefix= --compile --install-headers /home/sczone/.local/include/python3.8/horovod Check the logs for full command output.
```

```
    -- Found Pytorch: 1.9.1+cu111 (found suitable version "1.9.1+cu111", minimum required is "1.2.0")
    CMake Error at horovod/torch/CMakeLists.txt:18 (file):
      file failed to open for writing (Permission denied):
    
        /metadata.json
```
Hmm, something seems to have gone wrong here in such a way that `CMAKE_LIBRARY_OUTPUT_DIRECTORY_RELWITHDEBINFO` is not set to any value (`setup.py` handles that normally).

Probably it's related to this observation a bit further above, which seems to indicate that `cmake` is being rerun:
```
    You have changed variables that require your cache to be deleted.
    Configure will be re-run and you may have to reset some variables.
    The following variables have changed:
    CMAKE_CUDA_COMPILER= NOTFOUND
```
I don't really understand what could have happened here.

@you-old, if you can still reproduce the problem, could you try to add a `-v` to your `pip` command line and post the entire log? E.g.: `HOROVOD_WITH_PYTORCH=1 HOROVOD_WITHOUT_TENSORFLOW=1 pip3 install -v horovod[pytorch]`

----

@anthracene, for which frameworks did you need to manually install cmake before installing Horovod? Did you also get the message about re-running CMake Configure? In principle, it should work with the pinned `cmake~=3.13.0` -- we test that mechanism in one build configuration on the CI: https://github.com/horovod/horovod/blob/master/docker-compose.test.yml#L61

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -40,6 +40,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 - PyTorch: Fixed Reducescatter functions to raise `HorovodInternalError` rather than `RuntimeError`. ([#3594](https://github.com/horovod/horovod/pull/3594))
 - PyTorch on GPUs without GPU operations: Fixed grouped allreduce to set CPU device in tensor table. ([#3594](https://github.com/horovod/horovod/pull/3594))
 - Fixed race condition in PyTorch allocation handling. ([#3639](https://github.com/horovod/horovod/pull/3639))
+- Build: Fixed finding nvcc (if not in $PATH) with older versions of CMake. ([#3682](https://github.com/horovod/horovod/pull/3682))
 
 
 ## [v0.25.0] - 2022-06-20

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -163,6 +163,7 @@ if(NOT CMAKE_CUDA_COMPILER)
     find_package(CUDAToolkit)
     if(CUDAToolkit_BIN_DIR)
         message("CUDA compiler was not found in $PATH, but searching again in CUDA Toolkit binary directory")
+        unset(CMAKE_CUDA_COMPILER CACHE)  # need to clear this from cache, else some versions of CMake go into an infinite loop
         set(CMAKE_CUDA_COMPILER "${CUDAToolkit_BIN_DIR}/nvcc")
         check_language(CUDA)
     endif()

